File: solutions/matrix-diagonal-sum.py
# Matrix Diagonal Sum

from typing import List

def diagonal_sum(mat: List[List[int]]) -> int:
    # Sum diags row by row (0 + i, 0 - i)
    res = sum([mat[i][i] + mat[i][(len(mat) - 1) - i] for i in range(len(mat))])
    # If matrix has odd length, then center square was counted twice
    return res if len(mat) & 1 == 0 else res - mat[len(mat) // 2][len(mat) // 2] 

# Driver Code
mat = [[1,2,3],
       [4,5,6],
       [7,8,9]]
print(diagonal_sum(mat))

mat = [[1,1,1,1],
      [1,1,1,1],
      [1,1,1,1],
      [1,1,1,1]]
print(diagonal_sum(mat))

File: solutions/matrixDiagonalSum.js
// Matrix Diagonal Sum

const test = require('./test');

/**
 * Sum up diagonals, delete middle if matrix has odd number of rows 
 * since middle would be counted twice.
 *
 * @param {number[][]}
 * @return {number}
 */
const diagonalSum = mat => {
    // Sum row by row both diagonals (0 + i, length - i)
    let sum = mat.reduce((sum, row, i) => {
        return sum + row[i] + row[row.length - 1 - i];
    }, 0);
    // Calculate index of middle
    const mid = Math.floor(mat.length / 2);
    // Delete middle if matrix is odd
    return mat.length % 2 == 0 ? sum : sum - mat[mid][mid];
};

// Driver Code
const cases = [
    [[[1,2,3],
      [4,5,6],
      [7,8,9]], 25],
    [[[1,1,1,1],
      [1,1,1,1],
      [1,1,1,1],
      [1,1,1,1]], 8],
    [[], 0],
    [[[5]], 5]
];
test(cases, diagonalSum);
